include/RAJA/pattern/teams/teams_core.hpp

```diff
@@ -222,6 +224,25 @@ RAJA_HOST_DEVICE RAJA_INLINE void loop(CONTEXT const &ctx,
 #endif
 }
 
+template <typename POLICY_LIST,
+          typename CONTEXT,
+          typename SEGMENT,
+          typename BODY>
+RAJA_HOST_DEVICE RAJA_INLINE void loop_idx(CONTEXT const &ctx,
+                                          SEGMENT const &segment,
+                                          BODY const &body)
+{
+#if defined(RAJA_DEVICE_CODE)
+  LoopIdxExecute<typename POLICY_LIST::device_policy_t, SEGMENT>::exec(ctx,
+                                                                    segment,
+                                                                    body);
+#else
+  LoopIdxExecute<typename POLICY_LIST::host_policy_t, SEGMENT>::exec(ctx,
+                                                                  segment,
+                                                                  body);
+#endif
+}
+
 template <typename POLICY_LIST,
```

Comment:
loop_idx?

Reply:
that function returns the index within the tile, should we change the name?

Reply:
It does? It looks like it runs the loop?

Reply:
Ah yes, sorry I should be more clear, it does run the loop, the usage I had in mind was like this:

```
        RAJA::expt::loop_idx<loop_t>(ctx, r_tile, [&](int global_id, int loc_id) {

             }); // loop_idx
```
Where global_id is the the global index, and loc_id was the offset within the tile.

Reply:
Is that global_id of the tile, or global index in the dimension as a whole?  I ask partly because I'm not sure how useful the loc_id is if you don't have the tile_id as well, but the global index could be useful either way so I'm not even 100% sure which two I would want from this.

Reply:
It is the global index of the of iteration space as a whole. Currently, doing tiling within teams uses two loops. A tile loop (which loops over tiles), and the inner loop which loops over the values within the tile, usage looks like this: 

```
RAJA::expt::tile_idx<teams_x>(ctx, TILE_SZ, RAJA::RangeSegment(0, N_tri), [&](RAJA::RangeSegment const &r_tile, int tile_id) {           
  
  RAJA::expt::loop_idx<loop_t>(ctx, r_tile, [&](int global_id, int loc_id) {  

             }); // loop_idx                                                                                                                     
         });  // tile_idx         
```
In the tile_idx function, we loop over the tiles, r_tile is the tile that gets generated by partitioning the iteration space (and fed into loop_idx), and tile_id is the tile count.  Loop_idx then loops over the entries in r_tile, global_id is the global index with respect to the global iteration space, and loc_id is the local offset.

Reply:
It definitely depends on what is being coded up that determines what would be needed, these are just options to provide what may be needed. We also have variants that do not return the index (::tile<teams> , ::loop<loop_t>). I'm open to ideas, happy to get the discussion going.

Reply:
Ok, that makes more sense, thanks @artv3.  Is this also how the user expresses team/block and thread level parallelism or are these separate?

Reply:
Yes this is definitely one way, here users could map teams to tiles and threads to iterations within the tile. Shared memory can be allocated between the tile and loops so threads can share data.

Reply:
What's the other way? I suppose I was asking mainly to see if this is something where we could avoid some api duplication.

Reply:
The other way is to map teams/threads directly to loops, in finite element applications we have loops over elements, then inner loops over degree's of freedom inside an element ( No tiling necessary). 

In CUDA this would look like 

```
int e = blockIdx.x; 

     __shared__ double s_A[]; 

      int c = threadIdx.x 
      s_A[c] = .... 
```

Or using RAJA teams: 

```
RAJA::expt::loop<teams_x>(ctx, RAJA::RangeSegment(0, Elements), [&](int e) {

RAJA_SHARED double s_A[]

  RAJA::expt::loop<threads_x>(ctx, RAJA::RangeSegment(0, Ndofs), [&](int c) {
              s_A[c] = ... 
  }); 

});
```

The inner loops could be 1,2,3D and they would be doing a series of small matrix-matrix multiples.